File: Taller.POO_mini.Proyecto/Codigo en Java/Agenda_de_contactos.java
//agenda de contactos
//aqui es casi los mismo q en python esta clase representa un contacto general con tres datos básicos:
//nombre, teléfono y correo electrónico.

// 12.implementa en java las clases de tu diagrama.
class Contacto {
    protected String nombre;
    protected String telefono;
    protected String email;

    public Contacto(String nombre, String telefono, String email) {
        this.nombre = nombre;
        this.telefono = telefono;
        this.email = email;
    }

    // 14. muestra un ejemplo de polimorfismo en tu código.
    //polimorfismo:
    //aqui muestra la información general del contacto.
    public void mostrarInfo() {
        System.out.println("Nombre: " + nombre);
        System.out.println("Teléfono: " + telefono);
        System.out.println("Email: " + email);
    }
}


// 13. agrega atributos, métodos y una relación de herencia.
// Esta clase hereda de Contacto y agrega un nuevo dato osea el cumpleaños.
class ContactoAmigo extends Contacto {
    private String cumpleaños;  //aqui ps el atrivuto de los amiguitos

    //aqui llama al constructor de la clase padre con "super"
    // para inicializar nombre, teléfono y email,
    // y además inicializa el cumpleaños.
    public ContactoAmigo(String nombre, String telefono, String email, String cumpleaños) {
        super(nombre, telefono, email); 
        this.cumpleaños = cumpleaños;
    }
    //aqui se aplica el polimorfismo osea
    //se muestra primero la info basica (de Contacto)
    // y luego se agrega el dato específico del amigo (cumpleaños).
    @Override
    public void mostrarInfo() {
        super.mostrarInfo();  // Muestra los datos generales
        System.out.println("Cumpleaños: " + cumpleaños);
    }
}
//esta clase hereda de Contacto y agrega un nuevo dato: la empresa.
//representa un contacto laboral o profesional.
class ContactoTrabajo extends Contacto {
    private String empresa;  //este es el atributo propio de los contactos de trabajo

    //igual q antes use "super" para inicializar los datos básicos
    // y luego se asigna la empresa.
    public ContactoTrabajo(String nombre, String telefono, String email, String empresa) {
        super(nombre, telefono, email); 
        this.empresa = empresa;
    }
    @Override
    public void mostrarInfo() {
        super.mostrarInfo();
        System.out.println("Empresa: " + empresa);
    }
}


//esta clase contiene el método main(), osea punto de entrada del programa
public class Agenda_de_contactos {
    public static void main(String[] args) {

        //polimorfismo:
        //aqui es como hacer lo de tener los contactos pre definidos
        //los tengo clasificados en el contacto normal, el amigo, y el de trabajo 

        Contacto contacto1 = new Contacto("Valery Rincon", "310-123-4567", "valery@email.com");
        Contacto contacto2 = new ContactoAmigo("Luis Riveros", "320-987-6543", "luis@email.com", "15/07");
        Contacto contacto3 = new ContactoTrabajo("Paulo Gutierrez", "315-555-7890", "paulo@trabajo.com", "Diamante.SAS");
//utilice el "new" porq sirve para crear un objeto a partir de una clase, asi ya podia hacer los contactod



        // 15. ejecuta tu programa y muestra la salida.
        //aqui ya va la impresion de toda la agenda
        //(al inicio me quedo todo desordenado AJAJA no sabia q podia hacer en java esto tambien)

        System.out.println("=== Agenda de Contactos ===");
        contacto1.mostrarInfo();
        System.out.println("-------------------------");
        contacto2.mostrarInfo();
        System.out.println("-------------------------");
        contacto3.mostrarInfo(); 
        System.out.println("-------------------------");
        }    }

File: Taller.POO_mini.Proyecto/Codigo en Python/Agenda-de-contactos.py
#agenda de contactos
# la mini agenda mostrara los contactos (apenas 3 JAJA) que tiene como informacion:
# nombre, telefono y email.

# 12. implementa en python las clases de tu diagrama
# la clase base es contacto, que basicamente guarda lo mas basico:
# nombre, telefono y email.

class Contacto:
    def __init__(self, nombre, telefono, email):
        self.nombre = nombre
        self.telefono = telefono
        self.email = email

    def mostrar_info(self):
        print(f"nombre: {self.nombre}")
        print(f"telefono: {self.telefono}")
        print(f"email: {self.email}")


# luego aqui va la clase derivada contactoamigo
# 13. agrega atributos, metodos y relacion de herencia

class ContactoAmigo(Contacto):
    def __init__(self, nombre, telefono, email, cumpleaños):
        # super llama al constructor de la clase papa (contacto)
        # para no tener que escribir otra vez lo del nombre, telefono y email.
        super().__init__(nombre, telefono, email)
        self.cumpleaños = cumpleaños  # atributo nuevo solo de los amigos

    # sobrescribi el metodo mostrar_info (polimorfismo)
    def mostrar_info(self):
        super().mostrar_info()  # primero muestra los datos normales
        print(f"cumpleaños: {self.cumpleaños}")  # luego lo especial del amigo, que ps  en este caso es la fecha de cumple


# y aqui clase derivada contactotrabajo

class ContactoTrabajo(Contacto):
    def __init__(self, nombre, telefono, email, empresa):
        super().__init__(nombre, telefono, email)
        self.empresa = empresa  # atributo nuevo solo del trabajo

    def mostrar_info(self):
        super().mostrar_info()
        print(f"empresa: {self.empresa}")


# ---------------------------------------------------------------
# o sea: contactoamigo y contactotrabajo heredan de contacto
# esto significa que automaticamente tienen nombre, telefono y email
# sin tener que escribirlo otra vez. solo agregan lo que los hace distintos:
# contactoamigo → cumpleaños
# contactotrabajo → empresa
# pd: me embollé un poco en la herencia pero ya lo corregi jaja no te preocupes mayron.
# ---------------------------------------------------------------


# 14. ejemplo de polimorfismo
# aqui se demuestra el polimorfismo porque aunque todas las variables
# son tipo contacto, cada una se comporta diferente al ejecutar mostrar_info.
# cada clase imprime cosas distintas

# cree los contactos ya predefinidos, apenas 3 jaja
contacto1 = Contacto("valery rincon", 310-123-4567, "valery@email.com")   # valery como contacto normal
contacto2 = ContactoAmigo("luis riveros", 320-987-6543, "luis@email.com", "15/07")  # luis como contacto amigo
contacto3 = ContactoTrabajo("paulo gutierrez", 315-555-7890, "paulo@trabajo.com", "diamante.sas")  # paulo como contacto de trabajo


# 15. ejecucion del programa y salida
print("=== agenda de contactos ===")
contacto1.mostrar_info() 
print("-------------------------")
contacto2.mostrar_info()  
print("-------------------------")
contacto3.mostrar_info() 
print("-------------------------")
